testcases/crypto/aes_func.c

```diff
@@ -1929,6 +2006,308 @@ CK_RV do_SignVerifyMAC(struct published_mac_test_suite_info *tsuite)
+        testcase_fail("C_GetAttributeValue rc = %s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    testcase_pass("Generate AES key and change CKA_IBM_PROTKEY_EXTRACTABLE");
+
+testcase_cleanup:
+
+    if (h_key != CK_INVALID_HANDLE) {
+        rc = funcs->C_DestroyObject(session, h_key);
+        if (rc != CKR_OK)
+            testcase_error("C_DestroyObject rc=%s", p11_get_ckr(rc));
+    }
+
+    testcase_user_logout();
+    rc = funcs->C_CloseAllSessions(slot_id);
+    if (rc != CKR_OK)
+        testcase_error("C_CloseAllSessions rc=%s", p11_get_ckr(rc));
+
+    return rc;
+}
+
+CK_RV do_EncryptDecryptAESPkey()
+{
+    CK_ULONG keylen = 16;
+    CK_BBOOL btrue = CK_TRUE;
+    CK_BBOOL bfalse = CK_FALSE;
+    CK_RV rc = CKR_OK;
+    CK_MECHANISM keygen_mech = { CKM_AES_KEY_GEN, 0, 0 };
+    CK_MECHANISM encr_mech = { CKM_AES_ECB, 0, 0 };
+    CK_OBJECT_HANDLE h_key = CK_INVALID_HANDLE, h_key2 = CK_INVALID_HANDLE;
+    CK_SESSION_HANDLE session;
+    CK_UTF8CHAR label[] = "A test key for protected key support";
+    CK_ATTRIBUTE keygen_tmpl[] = {
+        {CKA_TOKEN, &btrue, sizeof(CK_BBOOL)},
+        {CKA_LABEL, &label, sizeof(label)},
+        {CKA_EXTRACTABLE, &bfalse, sizeof(CK_BBOOL)},
+        {CKA_VALUE_LEN, &keylen, sizeof(CK_ULONG)}
+    };
+    CK_ULONG keygen_tmpl_len = sizeof(keygen_tmpl) / sizeof(CK_ATTRIBUTE);
+
+    CK_ATTRIBUTE find_tmpl[] = {
+        {CKA_LABEL, &label, sizeof(label)},
+    };
+    CK_ULONG find_tmpl_len = sizeof(find_tmpl) / sizeof(CK_ATTRIBUTE);
+    CK_ULONG count = 0;
+
+    CK_BYTE input[] =
+        { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+          0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };
+    CK_BYTE output[32];
+    CK_BYTE decrypted[32];
+    CK_ULONG outlen, decrlen;
+
+    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN];
+    CK_FLAGS flags;
+    CK_ULONG user_pin_len;
+    CK_SLOT_ID slot_id = SLOT_ID;
+
+    testsuite_begin("do_EncryptDecryptAESPkey");
+    testcase_rw_session();
+    testcase_user_login();
+
+    /* Skip tests if the slot doesn't support protected keys*/
+    if (!is_ep11_token(slot_id)) {
+        testsuite_skip(6, "Slot %u doesn't support protected keys",
+                       (unsigned int) slot_id);
+        goto testcase_cleanup;
+    }
+
+    testcase_begin("Generate token key object and encrypt");
+    testcase_new_assertion();
+
+    /* Generate token object */
+    rc = funcs->C_GenerateKey(session, &keygen_mech, keygen_tmpl, keygen_tmpl_len, &h_key);
+    if (rc != CKR_OK) {
+        testcase_fail("C_GenerateKey rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    /* Encrypt some data with this key object */
+    rc = funcs->C_EncryptInit(session, &encr_mech, h_key);
+    if (rc != CKR_OK) {
+        testcase_fail("C_EncryptInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    memset(output, 0, sizeof(output));
+    rc = funcs->C_Encrypt(session, input, sizeof(input), output, &outlen);
```

Comment:
outlen must be initialized (to sizof(output)). 

I get the following when its uninitialzed:
```
* TESTCASE do_EncryptDecryptAESPkey BEGIN Generate token key object and encrypt
* TESTCASE do_EncryptDecryptAESPkey FAIL (testcases/crypto/aes_func.c:2172) C_Encrypt rc=CKR_BUFFER_TOO_SMALL

```

Reply:
Done.